rumqttc/src/client.rs

```diff
@@ -141,63 +141,91 @@ impl AsyncClient {
     where
         S: Into<String>,
     {
-        let mut publish = Publish::from_bytes(topic, qos, payload);
+        let topic = topic.into();
+        let mut publish = Publish::from_bytes(&topic, qos, payload);
         publish.retain = retain;
         let publish = Request::Publish(publish);
+        if !valid_topic(&topic) {
+            return Err(ClientError::Request(publish));
+        }
         self.request_tx.send_async(publish).await?;
         Ok(())
     }
 
     /// Sends a MQTT Subscribe to the `EventLoop`
     pub async fn subscribe<S: Into<String>>(&self, topic: S, qos: QoS) -> Result<(), ClientError> {
-        let subscribe = Subscribe::new(topic.into(), qos);
+        let topic = topic.into();
+        let subscribe = Subscribe::new(&topic, qos);
         let request = Request::Subscribe(subscribe);
+        if !valid_filter(topic) {
+            return Err(ClientError::Request(request));
+        }
         self.request_tx.send_async(request).await?;
         Ok(())
     }
 
     /// Attempts to send a MQTT Subscribe to the `EventLoop`
     pub fn try_subscribe<S: Into<String>>(&self, topic: S, qos: QoS) -> Result<(), ClientError> {
-        let subscribe = Subscribe::new(topic.into(), qos);
+        let topic = topic.into();
+        let subscribe = Subscribe::new(&topic, qos);
         let request = Request::Subscribe(subscribe);
+        if !valid_filter(topic) {
+            return Err(ClientError::Request(request));
+        }
         self.request_tx.try_send(request)?;
         Ok(())
     }
 
     /// Sends a MQTT Subscribe for multiple topics to the `EventLoop`
     pub async fn subscribe_many<T>(&self, topics: T) -> Result<(), ClientError>
     where
-        T: IntoIterator<Item = SubscribeFilter>,
+        T: IntoIterator<Item = SubscribeFilter> + Clone,
     {
+        let is_err = topics.clone().into_iter().any(|t| !valid_filter(t.path));
         let subscribe = Subscribe::new_many(topics);
         let request = Request::Subscribe(subscribe);
+        if is_err {
+            return Err(ClientError::Request(request));
+        }
         self.request_tx.send_async(request).await?;
         Ok(())
     }
 
     /// Attempts to send a MQTT Subscribe for multiple topics to the `EventLoop`
     pub fn try_subscribe_many<T>(&self, topics: T) -> Result<(), ClientError>
     where
-        T: IntoIterator<Item = SubscribeFilter>,
+        T: IntoIterator<Item = SubscribeFilter> + Clone,
     {
+        let is_err = topics.clone().into_iter().any(|t| !valid_filter(t.path));
```

Comment:
i think we don't need `clone` here since `valid_filter` only needs a reference.

Reply:
If `Clone` is not used then `into_iter()` would consume `topics`. Which means we won't be able to use it any further (eg, in argument to `Subscribe::new_many`).

Reply:
we can collect it early and iterate over its reference and pass it to below functions, i.e. functions below this will need to be changed from taking an iterator to taking `&[T]` or `T`